File: libs/core/src/declarations/declaration_table.cpp
/**
 * @file declaration_table.cpp
 * @author Francesco Corso
 * @date 8/15/2026
 * @brief
 */
module prism.core:declarations.declaration_table.impl;

import :declarations.declaration_table;
import :declarations.single_root_namespace_declaration;
import :declarations.visit;
import :semantic.compilation;

namespace prism
{
    class DeclarationTable::Cache final
    {
      public:
        explicit Cache(const DeclarationTable &table) : old_roots_{table.old_roots_}
        {
        }

        const MergedNamespaceDeclaration &merged_root() const
        {
            return *merged_root_.get_or_compute(
                [this]
                {
                    return MergedNamespaceDeclaration::create(
                        old_roots_ |
                        std::views::transform(
                            [](const LazyRootNamespace &lazy_root)
                            { return static_pointer_cast<const SingleNamespaceDeclaration>(lazy_root.get()); }) |
                        std::ranges::to<ImmutableArray>());
                });
        }

        const ImmutableHashSet<Name> &type_names() const
        {
            return type_names_.get_or_compute([this] { return get_type_names(merged_root()); });
        }

        const ImmutableHashSet<Name> &namespace_names() const
        {
            return namespace_names_.get_or_compute([this] { return get_namespace_names(merged_root()); });
        }

      private:
        ImmutableOrderedSet<LazyRootNamespace> old_roots_;
        mutable Lazy<RefCountPtr<const MergedNamespaceDeclaration>> merged_root_;
        mutable Lazy<ImmutableHashSet<Name>> type_names_;
        mutable Lazy<ImmutableHashSet<Name>> namespace_names_;
    };

    DeclarationTable::Builder::Builder(RefCountPtr<const DeclarationTable> table)
        : table_{table != nullptr ? std::move(table) : empty()}
    {
    }

    void DeclarationTable::Builder::add_root_declaration(LazyRootNamespace root)
    {
        realize_removes();
        added_roots_.push_back(std::move(root));
    }

    void DeclarationTable::Builder::remove_root_declaration(LazyRootNamespace root)
    {
        realize_adds();
        removed_roots_.push_back(std::move(root));
    }

    RefCountPtr<const DeclarationTable> DeclarationTable::Builder::build() &&
    {
        realize_adds();
        realize_removes();

        return std::move(table_);
    }

    void DeclarationTable::Builder::realize_adds()
    {
        if (added_roots_.empty())
            return;

        auto last_declaration = std::move(added_roots_.back());
        if (added_roots_.size() == 1)
        {
            if (!table_->latest_lazy_root_declaration_.has_value())
            {
                table_ = make_ref_counted<const DeclarationTable>(construct_tag,
                                                                  table_->old_roots_,
                                                                  std::move(last_declaration),
                                                                  table_->cache_);
            }
            else
            {
                table_ = make_ref_counted<const DeclarationTable>(
                    construct_tag,
                    table_->old_roots_.add(*table_->latest_lazy_root_declaration_),
                    std::move(last_declaration),
                    table_->cache_);
            }
        }
        else
        {
            added_roots_.pop_back();

            if (table_->latest_lazy_root_declaration_.has_value())
            {
                added_roots_.insert(added_roots_.begin(), *table_->latest_lazy_root_declaration_);
            }

            auto new_old_roots = table_->old_roots_.add_range(added_roots_ | std::views::as_rvalue);
        }

        added_roots_.clear();
    }

    void DeclarationTable::Builder::realize_removes()
    {
        if (removed_roots_.empty())
            return;

        if (removed_roots_.size() == 1)
        {
            auto first_declaration = std::move(removed_roots_.front());
            if (table_->latest_lazy_root_declaration_ == first_declaration)
            {
                table_ =
                    make_ref_counted<DeclarationTable>(construct_tag, table_->old_roots_, std::nullopt, table_->cache_);
            }
            else
            {
                table_ = make_ref_counted<DeclarationTable>(construct_tag,
                                                            table_->old_roots_.remove(first_declaration),
                                                            std::move(first_declaration),
                                                            table_->cache_);
            }
        }
        else
        {
            auto is_latest_removed = table_->latest_lazy_root_declaration_.has_value() &&
                                     std::ranges::contains(removed_roots_, *table_->latest_lazy_root_declaration_);

            auto new_old_roots = table_->old_roots_.remove_range(removed_roots_);
            auto new_latest = !is_latest_removed ? table_->latest_lazy_root_declaration_ : std::nullopt;

            table_ = make_ref_counted<DeclarationTable>(construct_tag,
                                                        std::move(new_old_roots),
                                                        std::move(new_latest),
                                                        table_->cache_);
        }

        removed_roots_.clear();
    }

    DeclarationTable::DeclarationTable(ConstructTag,
                                       ImmutableOrderedSet<LazyRootNamespace> old_roots,
                                       Optional<LazyRootNamespace> latest_lazy_root,
                                       std::shared_ptr<Cache> cache)
        : old_roots_{std::move(old_roots)}, latest_lazy_root_declaration_{std::move(latest_lazy_root)},
          cache_{cache != nullptr ? std::move(cache) : std::make_shared<Cache>(*this)}
    {
    }

    const RefCountPtr<const DeclarationTable> &DeclarationTable::empty()
    {
        static auto empty_table = make_ref_counted<const DeclarationTable>(construct_tag,
                                                                           ImmutableOrderedSet<LazyRootNamespace>{},
                                                                           std::nullopt,
                                                                           nullptr);
        return empty_table;
    }

    DeclarationTable::Builder DeclarationTable::to_builder() const
    {
        return Builder{shared_from_this()};
    }

    const MergedNamespaceDeclaration &DeclarationTable::get_merged_root(const Compilation &compilation) const
    {
        return *merged_root_.get_or_compute([&] { return calculate_merged_root(compilation); });
    }

    const ImmutableHashSet<Name> &DeclarationTable::type_names() const
    {
        return type_names_.get_or_compute([&] { return get_merged_type_names(); });
    }

    const ImmutableHashSet<Name> &DeclarationTable::namespace_names() const
    {
        return namespace_names_.get_or_compute([&] { return get_merged_namespace_names(); });
    }

    RefCountPtr<const MergedNamespaceDeclaration> DeclarationTable::calculate_merged_root(
        const Compilation &compilation) const
    {
        auto &old_root = cache_->merged_root();
        if (!latest_lazy_root_declaration_.has_value())
        {
            return old_root.shared_from_this();
        }

        auto &old_declarations = old_root.declarations();
        PooledVector<RefCountPtr<const SingleNamespaceDeclaration>> new_declarations;
        new_declarations.reserve(old_declarations.size() + 1);
        new_declarations.append_range(old_declarations);
        new_declarations.emplace_back(latest_lazy_root_declaration_->get());
        std::ranges::sort(new_declarations,
                          [&compilation](const RefCountPtr<const SingleNamespaceDeclaration> &lhs,
                                         const RefCountPtr<const SingleNamespaceDeclaration> &rhs)
                          {
                              return compilation.compare_source_locations(lhs->syntax_reference(),
                                                                          rhs->syntax_reference()) ==
                                     std::strong_ordering::less;
                          });
        return MergedNamespaceDeclaration::create(
            ImmutableArray{std::from_range, new_declarations | std::views::as_rvalue});
    }

    ImmutableHashSet<Name> DeclarationTable::get_merged_type_names() const
    {
        auto &cached_names = cache_->type_names();

        if (!latest_lazy_root_declaration_.has_value())
            return cached_names;

        return cached_names.add_range(get_type_names(*latest_lazy_root_declaration_->get()));
    }

    ImmutableHashSet<Name> DeclarationTable::get_merged_namespace_names() const
    {
        auto &cached_names = cache_->namespace_names();

        if (!latest_lazy_root_declaration_.has_value())
            return cached_names;

        return cached_names.add_range(get_namespace_names(*latest_lazy_root_declaration_->get()));
    }

    ImmutableHashSet<Name> DeclarationTable::get_type_names(const Declaration &declaration)
    {
        return get_names(declaration, [](const Declaration &d) { return d.kind() != DeclarationKind::namespace_; });
    }

    ImmutableHashSet<Name> DeclarationTable::get_namespace_names(const Declaration &declaration)
    {
        return get_names(declaration, [](const Declaration &d) { return d.kind() == DeclarationKind::namespace_; });
    }

    template <std::predicate<const Declaration &> Predicate>
    ImmutableHashSet<Name> DeclarationTable::get_names(const Declaration &declaration, Predicate &&predicate)
    {
        PooledSet<Name> set;
        PooledVector<Ref<const Declaration>> stack;
        stack.emplace_back(declaration);

        while (!stack.empty())
        {
            auto current = stack.back();
            stack.pop_back();

            if (std::invoke(predicate, current))
            {
                set.insert(current->name());
            }

            prism::visit(declaration,
                         [&](const auto &child)
                         {
                             for (const auto &member : child.members())
                             {
                                 stack.emplace_back(*member);
                             }
                         });
        }

        return ImmutableHashSet{std::from_range, set};
    }
} // namespace prism

File: src/Prism.Core/Declarations/DeclarationTable.cs
using System.Collections.Immutable;
using System.Runtime.InteropServices;
using Prism.Core.Compiling;
using Prism.Core.Utils;
using ZLinq;

namespace Prism.Core.Declarations;

internal sealed class DeclarationTable
{
    private readonly ImmutableOrderedSet<Lazy<SingleRootNamespaceDeclaration>> _oldRoots;
    private readonly Lazy<SingleRootNamespaceDeclaration>? _latestRoot;
    private Cache _cache;
    private MergedNamespaceDeclaration? _mergedRoot;

    private DeclarationTable(
        ImmutableOrderedSet<Lazy<SingleRootNamespaceDeclaration>> oldRoots,
        Lazy<SingleRootNamespaceDeclaration>? lazyRoot,
        Cache? cache
    )
    {
        _oldRoots = oldRoots;
        _latestRoot = lazyRoot;
        _cache = cache ?? new Cache(this);
    }

    public static readonly DeclarationTable Empty = new(
        ImmutableOrderedSet<Lazy<SingleRootNamespaceDeclaration>>.Empty,
        null,
        null
    );

    public Builder ToBuilder() => new(this);

    public MergedNamespaceDeclaration GetMergedRoot(Compilation compilation)
    {
        if (_mergedRoot is not null)
            return _mergedRoot;

        Interlocked.CompareExchange(ref _mergedRoot, CalculateMergedRoot(compilation), null);
        return _mergedRoot;
    }

    public ImmutableHashSet<string> TypeNames
    {
        get
        {
            if (field is not null)
                return field;

            Interlocked.CompareExchange(ref field, CalculateTypeNames(), null);
            return field;
        }
    }

    public ImmutableHashSet<string> NamespaceNames
    {
        get
        {
            if (field is not null)
                return field;

            Interlocked.CompareExchange(ref field, CalculateNamespaceNames(), null);
            return field;
        }
    }

    private MergedNamespaceDeclaration CalculateMergedRoot(Compilation compilation)
    {
        var oldRoot = _cache.MergedRoot;
        if (_latestRoot is null)
            return oldRoot;

        var oldDeclarations = oldRoot.Declarations;
        var newDeclarations = new SingleNamespaceDeclaration[oldDeclarations.Length + 1];
        oldDeclarations.CopyTo(newDeclarations);
        newDeclarations[^1] = _latestRoot.Value;
        Array.Sort(newDeclarations, new NamespaceLocationComparer(compilation));
        return MergedNamespaceDeclaration.Create(
            ImmutableCollectionsMarshal.AsImmutableArray(newDeclarations)
        );
    }

    private sealed class NamespaceLocationComparer(Compilation compilation)
        : IComparer<SingleNamespaceDeclaration>
    {
        public int Compare(SingleNamespaceDeclaration? x, SingleNamespaceDeclaration? y)
        {
            if (x is null)
            {
                return y is null ? 0 : 1;
            }

            if (y is null)
            {
                return -1;
            }

            return compilation.CompareSourceLocations(x.SyntaxReference, y.SyntaxReference);
        }
    }

    private ImmutableHashSet<string> CalculateTypeNames()
    {
        var cachedNames = _cache.TypeNames;

        return _latestRoot is null
            ? cachedNames
            : cachedNames.Union(GetTypeNames(_latestRoot.Value));
    }

    private ImmutableHashSet<string> CalculateNamespaceNames()
    {
        var cachedNames = _cache.TypeNames;

        return _latestRoot is null
            ? cachedNames
            : cachedNames.Union(GetNamespaceNames(_latestRoot.Value));
    }

    private static ImmutableHashSet<string> GetTypeNames(Declaration declaration)
    {
        return GetNames(declaration, d => d.Kind != DeclarationKind.Namespace);
    }

    private static ImmutableHashSet<string> GetNamespaceNames(Declaration declaration)
    {
        return GetNames(declaration, d => d.Kind != DeclarationKind.Namespace);
    }

    private static ImmutableHashSet<string> GetNames(
        Declaration declaration,
        Predicate<Declaration> predicate
    )
    {
        var builder = ImmutableHashSet.CreateBuilder<string>();
        var stack = new Stack<Declaration>();
        stack.Push(declaration);

        while (stack.TryPop(out var current))
        {
            if (predicate(current))
            {
                builder.Add(current.Name);
            }

            foreach (var member in declaration.Members)
            {
                stack.Push(member);
            }
        }

        return builder.ToImmutable();
    }

    private sealed class Cache
    {
        private readonly ImmutableOrderedSet<Lazy<SingleRootNamespaceDeclaration>> _oldRoots;

        public Cache(DeclarationTable table)
        {
            _oldRoots = table._oldRoots;
        }

        public MergedNamespaceDeclaration MergedRoot
        {
            get
            {
                if (field is not null)
                    return field;

                var merged = MergedNamespaceDeclaration.Create(
                    _oldRoots
                        .AsValueEnumerable()
                        .Select(SingleNamespaceDeclaration (x) => x.Value)
                        .ToImmutableArray()
                );
                Interlocked.CompareExchange(ref field, merged, null);
                return field;
            }
        }

        public ImmutableHashSet<string> TypeNames
        {
            get
            {
                if (field is not null)
                    return field;

                Interlocked.CompareExchange(ref field, GetTypeNames(MergedRoot), null);
                return field;
            }
        }

        public ImmutableHashSet<string> NamespaceNames
        {
            get
            {
                if (field is not null)
                    return field;

                Interlocked.CompareExchange(ref field, GetNamespaceNames(MergedRoot), null);
                return field;
            }
        }
    }

    public sealed class Builder(DeclarationTable? table)
    {
        private DeclarationTable _table = table ?? Empty;
        private List<Lazy<SingleRootNamespaceDeclaration>>? _pendingAdds;
        private List<Lazy<SingleRootNamespaceDeclaration>>? _pendingRemoves;

        public void AddRootDeclaration(Lazy<SingleRootNamespaceDeclaration> declaration)
        {
            RealizeRemoves();
            _pendingAdds ??= [];
            _pendingAdds.Add(declaration);
        }

        public void RemoveRootDeclaration(Lazy<SingleRootNamespaceDeclaration> declaration)
        {
            RealizeAdds();
            _pendingRemoves ??= [];
            _pendingRemoves.Add(declaration);
        }

        public DeclarationTable Build()
        {
            RealizeAdds();
            RealizeRemoves();
            return _table;
        }

        private void RealizeAdds()
        {
            if (_pendingAdds is null || _pendingAdds.Count == 0)
            {
                return;
            }

            var lastDeclaration = _pendingAdds[^1];
            if (_pendingAdds.Count == 1)
            {
                if (_table._latestRoot is null)
                {
                    _table = new DeclarationTable(_table._oldRoots, lastDeclaration, _table._cache);
                }
                else
                {
                    _table = new DeclarationTable(
                        _table._oldRoots.Add(_table._latestRoot),
                        lastDeclaration,
                        _table._cache
                    );
                }
            }
            else
            {
                _pendingAdds.RemoveAt(_pendingAdds.Count - 1);

                if (_table._latestRoot is not null)
                {
                    _pendingAdds.Insert(0, _table._latestRoot);
                }

                var newOldRoots = _table._oldRoots.AddRange(_pendingAdds);
                _table = new DeclarationTable(newOldRoots, lastDeclaration, cache: null);
            }

            _pendingAdds.Clear();
        }

        private void RealizeRemoves()
        {
            if (_pendingRemoves is null || _pendingRemoves.Count == 0)
                return;

            if (_pendingRemoves.Count == 1)
            {
                var firstDeclaration = _pendingRemoves[0];
                if (_table._latestRoot == firstDeclaration)
                {
                    _table = new DeclarationTable(_table._oldRoots, null, _table._cache);
                }
                else
                {
                    _table = new DeclarationTable(
                        _table._oldRoots.Remove(firstDeclaration),
                        firstDeclaration,
                        _table._cache
                    );
                }
            }
            else
            {
                var isLatestRemoved =
                    _table._latestRoot is not null && _pendingRemoves.Contains(_table._latestRoot);

                var newOldRoots = _table._oldRoots.RemoveRange(_pendingRemoves);
                var newLatest = !isLatestRemoved ? _table._latestRoot : null;

                _table = new DeclarationTable(newOldRoots, newLatest, _table._cache);
            }

            _pendingRemoves.Clear();
        }
    }
}
